pkgs/applications/networking/sync/celeste/default.nix

```
@@ -0,0 +1,94 @@
+{ lib
+, stdenv
+, rust
+, rustPlatform
+, fetchFromGitHub
+, substituteAll
+, fetchpatch
+, pkg-config
+, wrapGAppsHook4
+, cairo
+, gdk-pixbuf
+, glib
+, graphene
+, gtk3
+, gtk4
+, libadwaita
+, libclang
+, librclone
+, pango
+}:
+
+let
+  # https://github.com/trevyn/librclone/pull/8
+  librclone-mismatched-types-patch = fetchpatch {
+    name = "use-c_char-to-be-platform-independent.patch";
+    url = "https://github.com/trevyn/librclone/commit/91fdf3fa5f5eea0dfd06981ba72e09034974fdad.patch";
+    hash = "sha256-8YDyUNP/ISP5jCliT6UCxZ89fdRFud+6u6P29XdPy58=";
+  };
+in rustPlatform.buildRustPackage rec {
+  pname = "celeste";
+  version = "0.4.0";
+
+  src = fetchFromGitHub {
+    owner = "hwittenborn";
+    repo = "celeste";
+    rev = "v${version}";
+    hash = "sha256-6DUoqxVmj6TgnSLvyXNmmb2os/UtIO4O+f4NaG0xEWg=";
+  };
+
+  cargoHash = "sha256-3RtKr8ifDI3R1xZWqSH/ty1GOWN0b+IE2icDekVTW4A=";
+
+  patches = [
+    (substituteAll {
+      src = ./target-dir.patch;
+      rustTarget = rust.toRustTarget stdenv.hostPlatform;
+    })
+  ];
+
+  postPatch = ''
+    pushd $cargoDepsCopy/librclone-sys
+    oldHash=$(sha256sum build.rs | cut -d " " -f 1)
+    patch -p2 < ${./librclone-path.patch}
+    substituteInPlace build.rs \
+      --subst-var-by librclone ${librclone}
+    substituteInPlace .cargo-checksum.json \
+      --replace $oldHash $(sha256sum build.rs | cut -d " " -f 1)
+    popd
+    pushd $cargoDepsCopy/librclone
+    oldHash=$(sha256sum src/lib.rs | cut -d " " -f 1)
+    patch -p1 < ${librclone-mismatched-types-patch}
+    substituteInPlace .cargo-checksum.json \
+      --replace $oldHash $(sha256sum src/lib.rs | cut -d " " -f 1)
+    popd
+  '';
+
```

not the most idiomatic, but something like this could work

```suggestion
  cargoBuildFlags = [ "-p" "celeste-tray" ];
  postConfigure = ''
    cargoBuildHook
    cargoBuildFlags=
  '';

```

---

Wow, that's cursed. I'd also recommend adding a comment explaining why, and showing that [this is what upstream does to build it, too](https://github.com/hwittenborn/celeste/blob/765dfa25c5ce5a52bcaf347e94c8e8de5f27972b/justfile#L1-L3).

---

What's preferred? @figsoda's `-p` or upstreams `--bin`?

---

They're analogous, is the short answer. Maybe use upstream's just to maintain "consistency?"

---

Either one should be fine, use whichever you like